lib/topological_inventory/sync/iterator.rb

```diff
@@ -0,0 +1,29 @@
+module TopologicalInventory
+  class Sync
+    class Iterator
+      attr_reader :block, :limit, :resource_id
+
+      def initialize(block, limit: 100, resource_id: nil)
+        @block = block
+        @limit = limit
+        @resource_id = resource_id
+      end
+
+      def each
+        offset = 0
+
+        loop do
+          collection = resource_id.nil? ? block.call(limit, offset) : block.call(resource_id, limit, offset)
+          %i[data meta].each do |method|
+            raise "Sources API Client response expected" unless collection.respond_to?(method)
+          end
+          break if collection.data.blank?
+
+          collection.data.each { |record| yield record }
+          offset += limit
+          break if offset >= collection.meta.count.to_i
```

Comment:
Curious what others think here, but this guard would probably make sense at the beginning of the loop, it took me a bit to understand what was going on. Since this is (almost) a recursive function usually the base case is at the top, which this fits that pattern.

Reply:
the effect is probably the same. If you found it more readable on the top, I can move it

Reply:
..it depends on the response and has to be changed to `if offset + limit > collection.meta.count.to_i` and move maximally after the 'raise' block. 
Or how do you think it @lindgrenj6?

Reply:
oh yeah good point since it relies on the `collection`, I think after the raise block would be more readable, after the `break if collection.data.blank?` line.

Reply:
it'll skip the last round of yields. But I got the point and changed `loop` to `until finished`